src/probnum/linops/_linear_operator.py

```diff
@@ -129,19 +114,7 @@ def dot(self, x):
             1-d or 2-d array (depending on the shape of x) that represents
             the result of applying this linear operator on x.
         """
-        if isinstance(x, LinearOperator):
-            return _ProductLinearOperator(self, x)
-        elif np.isscalar(x):
-            return _ScaledLinearOperator(self, x)
-        else:
-            if len(x.shape) == 1 or len(x.shape) == 2 and x.shape[1] == 1:
-                return self.matvec(x)
-            elif len(x.shape) == 2:
-                return self.matmat(x)
-            else:
-                raise ValueError(
-                    "Expected 1-d or 2-d array, matrix or random variable, got %r." % x
-                )
+        return self @ x
```

Comment:
With this choice are we accepting possibly unexpected behavior when multiplying with a (m, n, d) array, right? In this scenario `np.matmul` treats the array as a stack of 2d arrays while `np.dot` treats it as a stack of vectors resulting in sets of inner products.

I am fine with this choice, but maybe we should raise an error here if anything but a 2d array is passed, like in the original implementation.

Reply:
I agree with your suggestion. In the next PR, I will remove the inheritance from `scipy.sparse.linalg.LinearOperator`. I would suggest to address this problem there, is that alright?

Reply:
I would be in favor of re-adding the `ValueError` already now, just so that in the meantime no accidental reliance on this behaviour is introduced, which then breaks in the next PR.

Reply:
Ah sorry, confused myself there 😄 The `ValueError` is still there, I just moved it to `__matmul__`